include/nbl/ext/ScreenShot/ScreenShot.h

```diff
@@ -18,228 +15,108 @@ namespace nbl
-					imgViewInfo.viewType = asset::IImageView<asset::ICPUImage>::ET_2D;
-					imgViewInfo.flags = static_cast<asset::ICPUImageView::E_CREATE_FLAGS>(0u);
-					imgViewInfo.subresourceRange.baseArrayLayer = 0u;
-					imgViewInfo.subresourceRange.baseMipLevel = 0u;
-					imgViewInfo.subresourceRange.layerCount = imgInfo.arrayLayers;
-					imgViewInfo.subresourceRange.levelCount = imgInfo.mipLevels;
+				video::IDriverMemoryBacked::SDriverMemoryRequirements driverMemoryRequirements;
+				core::smart_refctd_ptr<video::IGPUCommandBuffer> gpuCommandBuffer;
+				logicalDevice->createCommandBuffers(gpuCommandPool.get(), video::IGPUCommandBuffer::EL_PRIMARY, 1u, &gpuCommandBuffer);
+				assert(gpuCommandBuffer);
 
-					auto imageView = asset::ICPUImageView::create(std::move(imgViewInfo));
-					auto gpuImageView = driver->getGPUObjectsFromAssets(&imageView.get(), &imageView.get() + 1)->front();
+				nbl::core::smart_refctd_ptr<nbl::video::IGPUBuffer> gpuTexelBuffer;
 
-					return std::move(gpuImageView);
-				};
+				gpuCommandBuffer->begin(nbl::video::IGPUCommandBuffer::EU_ONE_TIME_SUBMIT_BIT);
+				{
+					auto extent = gpuImage->getMipSize();
+					video::IGPUImage::SBufferCopy pRegions[1u] = { {0u,extent.x,extent.y,{static_cast<asset::IImage::E_ASPECT_FLAGS>(0u),0,0u,1u},{0u,0u,0u},{extent.x,extent.y,extent.z}} };
 
-				auto gpuImageViewDepthBuffer = createAttachement(false);
-				auto gpuImageViewColorBuffer = createAttachement(true);
+					auto deviceLocalGPUMemoryReqs = logicalDevice->getDownStreamingMemoryReqs();
+					deviceLocalGPUMemoryReqs.vulkanReqs.size = extent.x * extent.y * extent.z * nbl::asset::getTexelOrBlockBytesize(fetchedGpuImageParams.format);
+					gpuTexelBuffer = logicalDevice->createGPUBufferOnDedMem(deviceLocalGPUMemoryReqs, true);
 
-				auto frameBuffer = driver->addFrameBuffer();
-				frameBuffer->attach(video::EFAP_DEPTH_ATTACHMENT, std::move(gpuImageViewDepthBuffer));
-				frameBuffer->attach(video::EFAP_COLOR_ATTACHMENT0, std::move(gpuImageViewColorBuffer));
+					gpuCommandBuffer->copyImageToBuffer(gpuImage.get(), nbl::asset::EIL_GENERAL, gpuTexelBuffer.get(), 1, pRegions);
+				}
+				gpuCommandBuffer->end();
 
-				return frameBuffer;
-			};
+				auto fence = logicalDevice->createFence(static_cast<nbl::video::IGPUFence::E_CREATE_FLAGS>(0));
 
-			/*
-				Download mip level image with gpu image usage and save it to IGPUBuffer.
-				Because of the fence placed by driver the function stalls the CPU 
-				to wait on the GPU to finish, beware of that.
-				@see video::IDriverFence
-			*/
+				video::IGPUQueue::SSubmitInfo info;
+				info.commandBufferCount = 1u;
+				info.commandBuffers = &gpuCommandBuffer.get();
+				info.pSignalSemaphores = nullptr;
+				info.signalSemaphoreCount = 0u;
+				info.pWaitSemaphores = &semaphore;
+				info.waitSemaphoreCount = 1u;
+				nbl::asset::E_PIPELINE_STAGE_FLAGS stageflags = nbl::asset::EPSF_TRANSFER_BIT;
+				info.pWaitDstStageMask = &stageflags;
```

Comment:
@Crisspl I think the stage flags are wrong, AFAIK he should wait for anything that could write to the texture he intends to copy?

Reply:
I decided to wait for everything